File: tests/test_parser.php
<?php
/**
 * test_parser.php — PHP HTTP Parser Test Client
 * Uses fsockopen for raw TCP. No cURL, no file_get_contents HTTP wrapper.
 *
 * Usage: php test_parser.php [host] [port]
 */

declare(strict_types=1);

$host = $argv[1] ?? '127.0.0.1';
$port = (int)($argv[2] ?? 8080);

$pass  = 0;
$fail  = 0;
$total = 0;

// ── helpers ───────────────────────────────────────────────────────────────────

function suite(string $name): void {
    echo "\n── {$name} ──\n";
}

function check(string $label, bool $condition, string $detail = ''): void {
    global $pass, $fail, $total;
    ++$total;
    if ($condition) {
        ++$pass;
        echo "  ✓  {$label}\n";
    } else {
        ++$fail;
        $d = $detail ? " ({$detail})" : '';
        echo "  ✗  FAIL: {$label}{$d}\n";
    }
}

/**
 * Open a TCP connection. Returns resource|false.
 * Timeout = 3 s for connect; 3 s for recv.
 */
function tcp_connect(string $host, int $port) {
    $errno  = 0;
    $errstr = '';
    $sock   = @fsockopen($host, $port, $errno, $errstr, 3.0);
    if ($sock === false) return false;
    stream_set_timeout($sock, 3);
    return $sock;
}

/** Read until timeout or EOF. */
function recv_all($sock): string {
    $data = '';
    while (!feof($sock)) {
        $chunk = fread($sock, 4096);
        if ($chunk === false || $chunk === '') {
            $info = stream_get_meta_data($sock);
            if ($info['timed_out']) break;
            break;
        }
        $data .= $chunk;
    }
    return $data;
}

/** Parse status code from raw HTTP response. */
function parse_status(string $raw): int {
    if (preg_match('/^HTTP\/\d\.\d (\d{3})/', $raw, $m)) {
        return (int)$m[1];
    }
    return 0;
}

/** Count occurrences of needle in haystack. */
function count_substr(string $haystack, string $needle): int {
    return substr_count($haystack, $needle);
}

// ── test cases ────────────────────────────────────────────────────────────────

function test_simple_get(string $host, int $port): void {
    suite('Simple GET');
    $s = tcp_connect($host, $port);
    if (!$s) { check('connect', false); return; }

    fwrite($s, "GET / HTTP/1.1\r\nHost: localhost\r\n\r\n");
    $resp = recv_all($s);
    fclose($s);

    $code = parse_status($resp);
    check('GET / returns 2xx or 3xx',  $code >= 200 && $code < 400, (string)$code);
    check('Response non-empty',         strlen($resp) > 0);
}

function test_post_with_body(string $host, int $port): void {
    suite('POST with body');
    $s = tcp_connect($host, $port);
    if (!$s) { check('connect', false); return; }

    $body = 'php_field=php_value&lang=php';
    $req  = "POST /echo HTTP/1.1\r\n"
          . "Host: localhost\r\n"
          . "Content-Type: application/x-www-form-urlencoded\r\n"
          . "Content-Length: " . strlen($body) . "\r\n"
          . "\r\n" . $body;

    fwrite($s, $req);
    $resp = recv_all($s);
    fclose($s);

    $code = parse_status($resp);
    check('POST — valid response code', $code > 0,   (string)$code);
    check('POST — not 5xx',             $code < 500, (string)$code);
}

function test_content_length_zero(string $host, int $port): void {
    suite('Content-Length: 0 (no body hang)');
    $s = tcp_connect($host, $port);
    if (!$s) { check('connect', false); return; }

    fwrite($s, "POST /empty HTTP/1.1\r\nHost: localhost\r\nContent-Length: 0\r\n\r\n");
    $resp = recv_all($s);
    fclose($s);

    check('CL:0 — server responds (no hang)',
          strlen($resp) > 0,
          'got no response — parser likely stuck in NEED_MORE');
    check('CL:0 — valid status code', parse_status($resp) > 0);
}

function test_leading_crlf(string $host, int $port): void {
    suite('Leading CRLF before request (telnet artifact)');
    $s = tcp_connect($host, $port);
    if (!$s) { check('connect', false); return; }

    fwrite($s, "\r\nGET /telnet-crlf HTTP/1.1\r\nHost: localhost\r\n\r\n");
    $resp = recv_all($s);
    fclose($s);

    $code = parse_status($resp);
    check('Leading CRLF — server responds',  strlen($resp) > 0);
    check('Leading CRLF — not 400',          $code !== 400, (string)$code);
}

function test_bad_request(string $host, int $port): void {
    suite('Malformed request → 400');
    $s = tcp_connect($host, $port);
    if (!$s) { check('connect', false); return; }

    fwrite($s, "GARBAGE NOT HTTP\r\n\r\n");
    $resp = recv_all($s);
    fclose($s);

    check('Bad request — returns 400', parse_status($resp) === 400, substr($resp, 0, 40));
}

function test_missing_host(string $host, int $port): void {
    suite('Missing Host header (HTTP/1.1) → 400');
    $s = tcp_connect($host, $port);
    if (!$s) { check('connect', false); return; }

    fwrite($s, "GET / HTTP/1.1\r\n\r\n");
    $resp = recv_all($s);
    fclose($s);

    check('Missing Host → 400', parse_status($resp) === 400, substr($resp, 0, 40));
}

function test_chunked_delivery(string $host, int $port): void {
    suite('Chunked delivery (8 bytes at a time)');
    $s = tcp_connect($host, $port);
    if (!$s) { check('connect', false); return; }

    $req  = "GET /chunked HTTP/1.1\r\nHost: localhost\r\n\r\n";
    $len  = strlen($req);
    for ($i = 0; $i < $len; $i += 8) {
        fwrite($s, substr($req, $i, 8));
        usleep(3000);   // 3 ms
    }

    $resp = recv_all($s);
    fclose($s);

    check('Chunked delivery — server responds',   strlen($resp) > 0);
    check('Chunked delivery — valid status code', parse_status($resp) > 0);
}

function test_keep_alive(string $host, int $port): void {
    suite('Keep-alive: 4 requests on one connection');
    $s = tcp_connect($host, $port);
    if (!$s) { check('connect', false); return; }

    for ($i = 1; $i <= 4; $i++) {
        $conn = ($i < 4) ? 'keep-alive' : 'close';
        fwrite($s,
            "GET /ka/php/{$i} HTTP/1.1\r\n"
            . "Host: localhost\r\n"
            . "Connection: {$conn}\r\n\r\n"
        );
    }

    $resp  = recv_all($s);
    fclose($s);
    $count = count_substr($resp, 'HTTP/1.');
    check('All 4 keep-alive responses received', $count === 4, "got {$count}");
}

function test_large_body(string $host, int $port): void {
    suite('Large body (64 KiB)');
    $s = tcp_connect($host, $port);
    if (!$s) { check('connect', false); return; }

    $body = str_repeat('P', 65536);
    $req  = "POST /large HTTP/1.1\r\n"
          . "Host: localhost\r\n"
          . "Content-Length: " . strlen($body) . "\r\n"
          . "\r\n" . $body;

    fwrite($s, $req);
    $resp = recv_all($s);
    fclose($s);

    $code = parse_status($resp);
    check('Large body — server responds', strlen($resp) > 0);
    check('Large body — not 5xx',         $code > 0 && $code < 500, (string)$code);
}

function test_pipelined(string $host, int $port): void {
    suite('Pipelined requests');
    $s = tcp_connect($host, $port);
    if (!$s) { check('connect', false); return; }

    fwrite($s,
        "GET /pipe/php/1 HTTP/1.1\r\nHost: localhost\r\n\r\n"
        . "GET /pipe/php/2 HTTP/1.1\r\nHost: localhost\r\n\r\n"
    );
    $resp  = recv_all($s);
    fclose($s);
    $count = count_substr($resp, 'HTTP/1.');
    check('Both pipelined responses received', $count === 2, "got {$count}");
}

function test_http10(string $host, int $port): void {
    suite('HTTP/1.0');
    $s = tcp_connect($host, $port);
    if (!$s) { check('connect', false); return; }

    fwrite($s, "GET / HTTP/1.0\r\n\r\n");
    $resp = recv_all($s);
    fclose($s);

    check('HTTP/1.0 — server responds',   strlen($resp) > 0);
    check('HTTP/1.0 — valid status code', parse_status($resp) > 0);
}

function test_content_length_variants(string $host, int $port): void {
    suite('Content-Length case variants');
    $variants = [
        'content-length', 'Content-Length', 'CONTENT-LENGTH', 'Content-length',
    ];
    foreach ($variants as $v) {
        $s = tcp_connect($host, $port);
        if (!$s) { check("connect ({$v})", false); continue; }

        $body = 'test';
        fwrite($s,
            "POST /case HTTP/1.1\r\nHost: localhost\r\n{$v}: " . strlen($body) . "\r\n\r\n{$body}"
        );
        $resp = recv_all($s);
        fclose($s);
        $code = parse_status($resp);
        check("CL variant '{$v}' — not 400", $code !== 400, (string)$code);
    }
}

function test_non_ascii_garbage(string $host, int $port): void {
    suite('Non-ASCII / binary garbage → 400');
    $s = tcp_connect($host, $port);
    if (!$s) { check('connect', false); return; }

    // Send raw binary bytes that are not valid HTTP
    fwrite($s, "\xff\xfe\x00\x01INVALID\r\n\r\n");
    $resp = recv_all($s);
    fclose($s);

    $code = parse_status($resp);
    check('Non-ASCII garbage — returns 400 or closes', $code === 400 || strlen($resp) === 0);
}

// ── main ──────────────────────────────────────────────────────────────────────

echo "PHP HTTP Parser Test Client → {$host}:{$port}\n";
echo str_repeat('=', 50) . "\n";

$probe = @fsockopen($host, $port, $errno, $errstr, 2.0);
if (!$probe) {
    fwrite(STDERR, "ERROR: Cannot connect to {$host}:{$port} — is the server running?\n");
    exit(1);
}
fclose($probe);

test_simple_get($host, $port);
test_post_with_body($host, $port);
test_content_length_zero($host, $port);
test_leading_crlf($host, $port);
test_bad_request($host, $port);
test_missing_host($host, $port);
test_chunked_delivery($host, $port);
test_keep_alive($host, $port);
test_large_body($host, $port);
test_pipelined($host, $port);
test_http10($host, $port);
test_content_length_variants($host, $port);
test_non_ascii_garbage($host, $port);

echo "\n" . str_repeat('=', 50) . "\n";
echo "  PHP results: {$pass}/{$total} passed";
if ($fail) echo "  ({$fail} failed)";
echo "\n" . str_repeat('=', 50) . "\n";

exit($fail > 0 ? 1 : 0);

File: tests/test_parser.rs
// test_parser.rs — Rust HTTP Parser Test Client
// Pure std::net::TcpStream. No reqwest, no hyper, no tokio.
//
// Build:  rustc -o test_parser_rs test_parser.rs
//   OR:   cargo run --bin test_parser -- [host] [port]
// Usage:  ./test_parser_rs [host] [port]

use std::io::{Read, Write};
use std::net::TcpStream;
use std::time::Duration;
use std::env;

// ── state ─────────────────────────────────────────────────────────────────────

struct Stats {
    pass:  u32,
    fail:  u32,
    total: u32,
}

impl Stats {
    fn new() -> Self { Stats { pass: 0, fail: 0, total: 0 } }

    fn check(&mut self, label: &str, condition: bool, detail: &str) {
        self.total += 1;
        if condition {
            self.pass += 1;
            println!("  ✓  {}", label);
        } else {
            self.fail += 1;
            if detail.is_empty() {
                println!("  ✗  FAIL: {}", label);
            } else {
                println!("  ✗  FAIL: {} ({})", label, &detail[..detail.len().min(80)]);
            }
        }
    }
}

// ── helpers ───────────────────────────────────────────────────────────────────

fn suite(name: &str) {
    println!("\n── {} ──", name);
}

fn connect(host: &str, port: u16) -> Option<TcpStream> {
    let addr = format!("{}:{}", host, port);
    match TcpStream::connect(&addr) {
        Ok(s) => {
            s.set_read_timeout(Some(Duration::from_secs(3))).ok();
            s.set_write_timeout(Some(Duration::from_secs(3))).ok();
            Some(s)
        }
        Err(e) => {
            eprintln!("  connect error: {}", e);
            None
        }
    }
}

fn send_all(stream: &mut TcpStream, data: &[u8]) -> bool {
    stream.write_all(data).is_ok()
}

fn recv_all(stream: &mut TcpStream) -> Vec<u8> {
    let mut buf  = Vec::new();
    let mut tmp  = [0u8; 4096];
    loop {
        match stream.read(&mut tmp) {
            Ok(0) => break,
            Ok(n) => buf.extend_from_slice(&tmp[..n]),
            Err(_) => break,
        }
    }
    buf
}

fn parse_status(raw: &[u8]) -> u16 {
    // "HTTP/1.x NNN ..."
    let s = std::str::from_utf8(raw).unwrap_or("");
    let sp = s.find(' ').unwrap_or(0);
    if sp == 0 { return 0; }
    s[sp+1..].splitn(2, ' ').next()
        .and_then(|t| t.trim().parse().ok())
        .unwrap_or(0)
}

fn count_occurrences(haystack: &[u8], needle: &[u8]) -> usize {
    let mut count = 0;
    let mut pos   = 0;
    while pos + needle.len() <= haystack.len() {
        if &haystack[pos..pos+needle.len()] == needle {
            count += 1;
            pos   += needle.len();
        } else {
            pos += 1;
        }
    }
    count
}

// ── test cases ────────────────────────────────────────────────────────────────

fn test_simple_get(st: &mut Stats, host: &str, port: u16) {
    suite("Simple GET");
    let mut s = match connect(host, port) {
        Some(s) => s,
        None    => { st.check("connect", false, ""); return; }
    };
    send_all(&mut s, b"GET / HTTP/1.1\r\nHost: localhost\r\n\r\n");
    let resp = recv_all(&mut s);
    let code = parse_status(&resp);
    st.check("GET / returns 2xx or 3xx", code >= 200 && code < 400, &code.to_string());
    st.check("Response non-empty",       !resp.is_empty(), "");
}

fn test_post_with_body(st: &mut Stats, host: &str, port: u16) {
    suite("POST with body");
    let mut s = match connect(host, port) {
        Some(s) => s,
        None    => { st.check("connect", false, ""); return; }
    };
    let body = b"rust_field=rust_value&version=2021";
    let req  = format!(
        "POST /echo HTTP/1.1\r\nHost: localhost\r\n\
         Content-Type: application/x-www-form-urlencoded\r\n\
         Content-Length: {}\r\n\r\n",
        body.len()
    );
    send_all(&mut s, req.as_bytes());
    send_all(&mut s, body);
    let resp = recv_all(&mut s);
    let code = parse_status(&resp);
    st.check("POST — valid response code", code > 0,   &code.to_string());
    st.check("POST — not 5xx",             code < 500, &code.to_string());
}

fn test_content_length_zero(st: &mut Stats, host: &str, port: u16) {
    suite("Content-Length: 0 (no body hang)");
    let mut s = match connect(host, port) {
        Some(s) => s,
        None    => { st.check("connect", false, ""); return; }
    };
    send_all(&mut s, b"POST /empty HTTP/1.1\r\nHost: localhost\r\nContent-Length: 0\r\n\r\n");
    let resp = recv_all(&mut s);
    st.check(
        "CL:0 — server responds (no hang)",
        !resp.is_empty(),
        "got no response — parser likely stuck in NEED_MORE",
    );
    st.check("CL:0 — valid status code", parse_status(&resp) > 0, "");
}

fn test_leading_crlf(st: &mut Stats, host: &str, port: u16) {
    suite("Leading CRLF before request (telnet artifact)");
    let mut s = match connect(host, port) {
        Some(s) => s,
        None    => { st.check("connect", false, ""); return; }
    };
    send_all(&mut s, b"\r\nGET /telnet-crlf HTTP/1.1\r\nHost: localhost\r\n\r\n");
    let resp = recv_all(&mut s);
    let code = parse_status(&resp);
    st.check("Leading CRLF — server responds", !resp.is_empty(), "");
    st.check("Leading CRLF — not 400",         code != 400, &code.to_string());
}

fn test_bad_request(st: &mut Stats, host: &str, port: u16) {
    suite("Malformed request → 400");
    let mut s = match connect(host, port) {
        Some(s) => s,
        None    => { st.check("connect", false, ""); return; }
    };
    send_all(&mut s, b"GARBAGE NOT HTTP AT ALL\r\n\r\n");
    let resp = recv_all(&mut s);
    let code = parse_status(&resp);
    st.check("Bad request — returns 400", code == 400, &code.to_string());
}

fn test_missing_host(st: &mut Stats, host: &str, port: u16) {
    suite("Missing Host header (HTTP/1.1) → 400");
    let mut s = match connect(host, port) {
        Some(s) => s,
        None    => { st.check("connect", false, ""); return; }
    };
    send_all(&mut s, b"GET / HTTP/1.1\r\n\r\n");
    let resp = recv_all(&mut s);
    let code = parse_status(&resp);
    st.check("Missing Host → 400", code == 400, &code.to_string());
}

fn test_byte_by_byte(st: &mut Stats, host: &str, port: u16) {
    suite("Byte-by-byte delivery (parser accumulation)");
    let mut s = match connect(host, port) {
        Some(s) => s,
        None    => { st.check("connect", false, ""); return; }
    };
    let req = b"GET /bytewise HTTP/1.1\r\nHost: localhost\r\n\r\n";
    for byte in req.iter() {
        if s.write_all(&[*byte]).is_err() { break; }
        std::thread::sleep(Duration::from_micros(500));
    }
    let resp = recv_all(&mut s);
    st.check("Byte-by-byte — server responds",   !resp.is_empty(), "");
    st.check("Byte-by-byte — valid status code", parse_status(&resp) > 0, "");
}

fn test_large_body(st: &mut Stats, host: &str, port: u16) {
    suite("Large body (64 KiB)");
    let mut s = match connect(host, port) {
        Some(s) => s,
        None    => { st.check("connect", false, ""); return; }
    };
    let body = vec![b'R'; 65536];
    let header = format!(
        "POST /large HTTP/1.1\r\nHost: localhost\r\nContent-Length: {}\r\n\r\n",
        body.len()
    );
    send_all(&mut s, header.as_bytes());
    send_all(&mut s, &body);
    let resp = recv_all(&mut s);
    let code = parse_status(&resp);
    st.check("Large body — server responds", !resp.is_empty(), "");
    st.check("Large body — not 5xx",         code > 0 && code < 500, &code.to_string());
}

fn test_pipelined(st: &mut Stats, host: &str, port: u16) {
    suite("Pipelined requests");
    let mut s = match connect(host, port) {
        Some(s) => s,
        None    => { st.check("connect", false, ""); return; }
    };
    send_all(&mut s,
        b"GET /pipe/rs/1 HTTP/1.1\r\nHost: localhost\r\n\r\n\
          GET /pipe/rs/2 HTTP/1.1\r\nHost: localhost\r\n\r\n"
    );
    let resp  = recv_all(&mut s);
    let count = count_occurrences(&resp, b"HTTP/1.");
    st.check("Both pipelined responses received", count == 2, &format!("got {}", count));
}

fn test_http10(st: &mut Stats, host: &str, port: u16) {
    suite("HTTP/1.0");
    let mut s = match connect(host, port) {
        Some(s) => s,
        None    => { st.check("connect", false, ""); return; }
    };
    send_all(&mut s, b"GET / HTTP/1.0\r\n\r\n");
    let resp = recv_all(&mut s);
    st.check("HTTP/1.0 — server responds",   !resp.is_empty(), "");
    st.check("HTTP/1.0 — valid status code", parse_status(&resp) > 0, "");
}

fn test_keep_alive(st: &mut Stats, host: &str, port: u16) {
    suite("Keep-alive: 5 requests on one connection");
    let mut s = match connect(host, port) {
        Some(s) => s,
        None    => { st.check("connect", false, ""); return; }
    };
    for i in 1..=5u32 {
        let conn = if i < 5 { "keep-alive" } else { "close" };
        let req  = format!(
            "GET /ka/rs/{} HTTP/1.1\r\nHost: localhost\r\nConnection: {}\r\n\r\n",
            i, conn
        );
        if send_all(&mut s, req.as_bytes()) { } else { break; }
    }
    let resp  = recv_all(&mut s);
    let count = count_occurrences(&resp, b"HTTP/1.");
    st.check("All 5 keep-alive responses received", count == 5, &format!("got {}", count));
}

fn test_header_case_variants(st: &mut Stats, host: &str, port: u16) {
    suite("Content-Length case variants");
    let variants = [
        "content-length", "Content-Length", "CONTENT-LENGTH", "cOnTeNt-LeNgTh",
    ];
    for v in &variants {
        let mut s = match connect(host, port) {
            Some(s) => s,
            None    => { st.check(&format!("connect ({})", v), false, ""); continue; }
        };
        let body = "hi";
        let req  = format!(
            "POST /case HTTP/1.1\r\nHost: localhost\r\n{}: {}\r\n\r\n{}",
            v, body.len(), body
        );
        send_all(&mut s, req.as_bytes());
        let resp = recv_all(&mut s);
        let code = parse_status(&resp);
        st.check(
            &format!("CL variant '{}' — not 400", v),
            code != 400,
            &code.to_string(),
        );
    }
}

fn test_non_ascii_garbage(st: &mut Stats, host: &str, port: u16) {
    suite("Non-ASCII / binary garbage → 400");
    let mut s = match connect(host, port) {
        Some(s) => s,
        None    => { st.check("connect", false, ""); return; }
    };
    send_all(&mut s, b"\xff\xfe\xfd INVALID HTTP\r\n\r\n");
    let resp = recv_all(&mut s);
    let code = parse_status(&resp);
    st.check(
        "Binary garbage — returns 400 or closes",
        code == 400 || resp.is_empty(),
        &code.to_string(),
    );
}

// ── main ──────────────────────────────────────────────────────────────────────

fn main() {
    let args: Vec<String> = env::args().collect();
    let host = args.get(1).map(|s| s.as_str()).unwrap_or("127.0.0.1");
    let port: u16 = args.get(2).and_then(|s| s.parse().ok()).unwrap_or(8080);

    println!("Rust HTTP Parser Test Client → {}:{}", host, port);
    println!("{}", "=".repeat(50));

    // Connectivity check
    if connect(host, port).is_none() {
        eprintln!("ERROR: Cannot connect to {}:{} — is the server running?", host, port);
        std::process::exit(1);
    }

    let mut st = Stats::new();

    test_simple_get          (&mut st, host, port);
    test_post_with_body      (&mut st, host, port);
    test_content_length_zero (&mut st, host, port);
    test_leading_crlf        (&mut st, host, port);
    test_bad_request         (&mut st, host, port);
    test_missing_host        (&mut st, host, port);
    test_byte_by_byte        (&mut st, host, port);
    test_large_body          (&mut st, host, port);
    test_pipelined           (&mut st, host, port);
    test_http10              (&mut st, host, port);
    test_keep_alive          (&mut st, host, port);
    test_header_case_variants(&mut st, host, port);
    test_non_ascii_garbage   (&mut st, host, port);

    println!("\n{}", "=".repeat(50));
    print!("  Rust results: {}/{} passed", st.pass, st.total);
    if st.fail > 0 { print!("  ({} failed)", st.fail); }
    println!("\n{}", "=".repeat(50));

    std::process::exit(if st.fail > 0 { 1 } else { 0 });
}
